react/lesson1.md

```diff
@@ -36,175 +91,431 @@ cd my-app/
   }
-}
+```
 
-export default App;
+Modify the `buildGame()` function so it only selects the button that's been marked on the state:
 
+```javascript
+buttons.push(<GameButton selected={this.state.selectedButton === i}/>)
 ```
 
-- [State Exercise](https://codepen.io/kabaros/pen/NdRNmm?editors=0010#0)
+# Add a timer
+
+Let's a delay between the time we press "Start game" and one of the buttons getting selected:
+
+```javascript
+  startGame = () => {
+    setTimeout(this.selectRandomButton, 2000);
+  }
+
+  ///render()
+  <button onClick={this.startGame}>Start game</button> 
+```
+
+# Add a click handler to our buttons
+
+```javascript
+<GameButton onClick={this.gameButtonClick} selected={this.state.selectedButton === i}/>
+```
+
+And our click handler:
+
+```javascript
+gameButtonClick = (event) => {
+  const className = event.target.className;
+  if(className.includes('selected')) {
+    alert("GOOD JOB");
+  }
+};
+```
+
+Click on "Start game" and wait for one of the buttons to become selected. Once a button becomes selected, click on it. Hmm, it still doesn't work.
+
+That's because inside our `GameButton` component, we are doing nothing to handle the `onClick` prop we're sending in. Let's change that!
+
+Inside `gameButton.js`:
+
+```javascript
+<button onClick={props.onClick}...
+```
+
+# Calculate how fast you click on a button
+
+This is the point where we add a challenge and turn our app into a game. We need to calculate how fast the user reacts to a button being selected. 
+
+Add a start and end time to the state:
+
+```javascript
+  constructor() {
+    super();
+    this.state = {
+      difficulty: 10,
+      selectedButton: -1,
+      startTime: 0
+    };
+  }
+
+  // When a random gets selected set the START TIME
+  selectRandomButton = () => {
+    const randomIndex = this.getRandomInt(0, this.state.difficulty);
+    this.setState({
+      selectedButton: randomIndex,
+      startTime: new Date().getTime()
+    })
+  }
+
+  // When the user manages to click on the selected button, display the time passed start and end
+  gameButtonClick = (event) => {
+    const className = event.target.className;
+    if(className.includes('selected')) {
+      const currTime = new Date().getTime();
+      alert(currTime - this.state.startTime);
+    }
+  };
+```
+
+# Instead of alerting the time, display it on the screen
+
+Add an `endTime` property to the state, initialized with 0.
+
+Modify you game button click function so that it stores the time it was clicked on the state (and remove the alert):
+
+```javascript
+gameButtonClick = (event) => {
+  const className = event.target.className;
+  if(className.includes('selected')) {
+    this.setState({
+      endTime: new Date().getTime() 
+    });
+  }
+};
+```
+
+In your render function, print out the final value:
+
+```javascript
+Score: {
+  this.state.endTime > 0
+    ? this.state.endTime - this.state.startTime
+    : "Not yet started.."
+}
+```
```

Comment:
can this ultimately get moved to a function outside of render? `displayScore` or something.

Reply:
It can. Although I mainly wanted to show an example of using a ternary operator in JSX. Don't mind moving it out though if you think it's better?

Reply:
I think we've already got a ternary above this, I find multi-line ternaries are quite hard to read when compared against a function with an `if`/`else`

Reply:
I've gotten used to ternaries, especially when on multiple lines, sometimes I find them much easier to read. But the students are definitely more used to a simple if/else, so I'll move it out, as you suggested.